File: php/day7_part2_2020.php
<?php
$file = fopen("input.txt", "r") or die("Unable to open file");

$rules = array();
$ruleRegex = "/(\d+) (\w+ \w+) bags?[,.]*/";

while(!feof($file)) {
    $line = fgets($file);
    $parts = explode(" bags contain ", $line);
    $container = $parts[0];
    $contents = $parts[1];

    if ($contents == "no other bags.") {
        continue;
    }

    preg_match_all($ruleRegex, $contents, $matches, PREG_SET_ORDER);
    foreach ($matches as $match) {
        $count = intval($match[1]);
        $rules[$container][] = array("color" => $match[2], "count" => $count);
    }
}
fclose($file);

$totalBags = countBags("shiny gold", $rules) - 1;
echo $totalBags;

function countBags($color, $rules) {
    $count = 1;
    if (array_key_exists($color, $rules)) {
        foreach ($rules[$color] as $rule) {
            $count += $rule["count"] * countBags($rule["color"], $rules);
        }
    }
    return $count;
}
?>

File: matlab/day7_part2_2020.m

function main()
    rules = containers.Map('KeyType', 'char', 'ValueType', 'any');
    fileId = fopen('input.txt', 'r');
    
    while ~feof(fileId)
        line = strtrim(fgets(fileId));
        parts = strsplit(line, ' bags contain ');
        container = parts{1};
        contents = parts{2};
        
        if strcmp(contents, 'no other bags.')
            continue;
        end
        
        bagRules = {};
        matches = regexp(contents, '(\d+) (\w+ \w+) bags?[,.]', 'tokens');
        
        for i = 1:length(matches)
            count = str2double(matches{i}{1});
            color = matches{i}{2};
            bagRules{end+1} = struct('Color', color, 'Count', count);
        end
        
        rules(container) = bagRules;
    end
    
    fclose(fileId);
    
    totalBags = countBags('shiny gold', rules) - 1;
    disp(totalBags);
end

function count = countBags(color, rules)
    count = 1;
    if isKey(rules, color)
        bagRules = rules(color);
        for i = 1:length(bagRules)
            count = count + bagRules{i}.Count * countBags(bagRules{i}.Color, rules);
        end
    end
end

main();
